src/services/balances.rs

```diff
@@ -43,28 +47,90 @@ pub async fn balances(
 
     let mut total_fiat = 0.0;
 
-    let mut service_balances: Vec<Balance> = backend_balances
-        .into_iter()
+    let token_addresses = backend_balances
+        .iter()
+        .map(|balance| {
+            balance
+                .token_address
+                .to_owned()
+                .unwrap_or("0x0000000000000000000000000000000000000000".to_string())
+        })
+        .collect::<Vec<_>>();
+
+    // We collect the TokenPrice which were successful – unsuccessful ones are ignored
+    let token_prices: Vec<TokenPrice> = stream::iter(token_addresses)
+        .map(|token_address| get_token_usd_rate(context, chain_id, token_address))
+        .buffer_unordered(N_CONCURRENT_TOKEN_REQUESTS)
+        .filter_map(|t| async move {
+            match t {
+                Ok(token_price) => Some(token_price),
+                Err(_) => None,
+            }
+        })
+        .collect::<Vec<_>>()
+        .await;
+
+    let service_balances: Vec<Balance> = backend_balances
+        .iter()
         .map(|it| {
-            let balance = it.to_balance(usd_to_fiat, &native_currency);
+            let token_address: String = it
+                .token_address
+                .to_owned()
+                .unwrap_or("0x0000000000000000000000000000000000000000".to_string());
+            let token_price: Option<&TokenPrice> = token_prices
+                .iter()
+                .find(|&token_price| token_price.address == token_address);
+            let token_to_usd: f64 = token_price
+                .and_then(|t| Some(t.fiat_price))
+                .unwrap_or(f64::from(0));
+
+            let balance = it.to_balance(token_to_usd, usd_to_fiat, &native_currency);
             total_fiat += balance.fiat_balance.parse::<f64>().unwrap_or(0.0);
             balance
         })
         .collect();
 
-    service_balances.sort_by(|a, b| {
-        b.fiat_balance
-            .parse::<f64>()
-            .unwrap_or(0.0)
-            .partial_cmp(&a.fiat_balance.parse::<f64>().unwrap_or(0.0))
-            .unwrap_or(Ordering::Equal)
-    });
     Ok(Balances {
         fiat_total: total_fiat.to_string(),
         items: service_balances,
     })
 }
 
+/// Gets the [TokenPrice] of the token with address [token_address] for the chain [chain_id]
+/// To retrieve the Native Currency fiat price of the chain (eg.: Ether), 0x0000000000000000000000000000000000000000 should be used
+///
+/// # Arguments
+///
+/// * `context`: The context where the request will be executed
+/// * `chain_id`: The chain id on which this request should be executed
+/// * `token_address`: The token address (0x0000000000000000000000000000000000000000 for native currency)
+///
+/// returns: Result<TokenPrice, ApiError>
+///
```

Comment:
the module `services` is hidden from the cargo documentation generator. Documentation, such as this one is normally placed at the `route` module level.

Reply:
I thought this could still be used by editors/IDEs or even just for reading purposes. Eg.: IntelliJ generates the following:
![image](https://user-images.githubusercontent.com/3332770/135289952-91d5a27d-ed2e-4dbf-baf2-aad73107f9ab.png)

But I can remove it 👍

Reply:
You can just move it to `routes/balances` :) Or just adjust the ones there, as those are the ones that get pushed to the docs in github pages